logicnet/validator/challenger/challenger.py

```diff
@@ -1,53 +1,153 @@
 # Challenge for Synthetic Request
+import os
 import openai
 import random
 from logicnet.protocol import LogicSynapse
 import bittensor as bt
 from .human_noise import get_condition
 from .math_generator.topics import TOPICS as topics
 import mathgenerator
+from datasets import load_dataset
 
+DATASET_WEIGHT = [40,10,10,10,10,10,10]
 
 class LogicChallenger:
-    def __init__(self, base_url: str, api_key: str, model: str):
+    def __init__(self, base_url: str, api_key: str, model: str, dataset_weight: list):
         bt.logging.info(
-            f"Logic Challenger initialized with model: {model}, base_url: {base_url}"
+            f"Initializing Logic Challenger with model: {model}, base URL: {base_url}."
         )
         self.model = model
         self.openai_client = openai.OpenAI(base_url=base_url, api_key=api_key)
+        self.dataset_weight = [float(weight) for weight in dataset_weight.split(',')]
 
     def __call__(self, synapse: LogicSynapse) -> LogicSynapse:
         self.get_challenge(synapse)
         return synapse
 
     def get_challenge(self, synapse: LogicSynapse):
-        logic_problem = self.get_atom_math_problem(synapse)
+        # Get an atom logic problem
+        atom_logic_question, atom_logic_answer = self.get_atom_logic_problem()
+        if atom_logic_question is None or atom_logic_answer is None:
+            bt.logging.error("Unable to retrieve atom logic problem. Retrying...")
+            atom_logic_question, atom_logic_answer = self.get_atom_logic_problem()
+
+        # Revise the problem
         conditions: dict = get_condition()
-        revised_logic_question: str = self.get_revised_math_question(
-            logic_problem, conditions
+        revised_logic_question: str = self.get_revised_logic_question(
+            atom_logic_question, conditions
         )
+        
+        # Set the synapse with the atom problem
+        synapse.raw_logic_question = atom_logic_question
+        synapse.ground_truth_answer = str(atom_logic_answer).replace("$", "").strip()
```

Comment:
This line have problem with new datasets, some of the new datasets have proper human-like answers and not just number or symbols so it kinda messed up the `ground_truth_answer` like this for example:

- When using mathgenerator as question dataset for question generator we have good ground truth like this:
    
    ```bash
    0|sn35-validator-openai  | 2024-12-08 04:42:02.984 |      DEBUG       | [CORRECTNESS] Ground truth: 4.97
    0|sn35-validator-openai  | 2024-12-08 04:42:02.989 |      DEBUG       | [CORRECTNESS] Used programmatic comparison for response 0 with score 1.00000000000000 for answer $4.97$ against ground truth 4.97
    
    ```
    
- But when using different datasets, some datasets got this problem:
    
    ```bash
    0|sn35-validator-openai  | 2024-12-08 04:43:12.679 |      DEBUG       | [CORRECTNESS] Ground truth: The height of the sunflowers from Packet A is the same as the height of the sunflowers from Packet B + 20%. If Y represents the height of the sunflowers from Packet B, Y + 0.20Y = the height of sunflowers from Packet A.
    0|sn35-validator-openai  | We know the height of the sunflowers from Packet A is 192 inches = Y + 0.20Y or 1.2Y.
    0|sn35-validator-openai  | To figure out Y, we would divide both sides of the equation by 1.2, like this: 192 / 1.2 = 1.2Y / 1.2 or 160 = Y.
    0|sn35-validator-openai  | #### 160
    0|sn35-validator-openai  | 2024-12-08 04:43:12.683 |      DEBUG       | [CORRECTNESS] Unable to use programmatic comparison. Need LLM evaluation for response 0 with answer \( h = 160 \) inches. against ground truth The height of the sunflowers from Packet A is the same as the height of the sunflowers from Packet B + 20%. If Y represents the height of the sunflowers from Packet B, Y + 0.20Y = the height of sunflowers from Packet A.
    0|sn35-validator-openai  | We know the height of the sunflowers from Packet A is 192 inches = Y + 0.20Y or 1.2Y.
    0|sn35-validator-openai  | To figure out Y, we would divide both sides of the equation by 1.2, like this: 192 / 1.2 = 1.2Y / 1.2 or 160 = Y.
    0|sn35-validator-openai  | #### 160
    ```
    

So we need a better `ground_truth_answer` extraction purely for the correctness evaluation, I suggest just use the running LLM with some prompt engineering instuction to do that

Reply:
There are several datasets available for inclusion in the data pool, and the `ground_truth_answer` can adopt various formats depending on the dataset.

To enable accurate programmatic comparison, it is essential to fine-tune the LLM specifically for extraction tasks. This ensures that the AI generates a reasonable `ground_truth_answer` based on the raw answer. However, due to the diversity of datasets, the un-tuned LLM may fail to produce a valid `ground_truth_answer`, making programmatic comparisons unreliable. Consequently, this could lead to incorrect or unreasonable answers, resulting in unfair scoring for miner responses, especially since the AI layer is integrated into the scoring process. This presents a potential drawback in our subnet.

In the absence of such fine-tuning, we would need to rely solely on LLM-based evaluation for miner responses. While this method may offer a fairer approach compared to an unreliable programmatic method, fine-tuning the LLM remains a key strategy to address these issues and achieve our goal of implementing a robust programmatic scoring mechanism.